packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol

```diff
@@ -0,0 +1,1167 @@
+    ///         new proxy instead. Calls to source contracts MUST NOT fail under any circumstances
+    ///         other than the function not existing (which can happen in an upgrade scenario).
+    /// @param _source The source contract to load the proxy from.
+    /// @param _selector The selector of the function to call on the source contract.
+    /// @param _args The basic arguments for the proxy deployment.
+    /// @param _contractName The name of the contract to deploy.
+    /// @param _instructions The extra upgrade instructions for the proxy deployment.
+    /// @return The address of the loaded or built proxy.
+    function _loadOrDeployProxy(
+        address _source,
+        bytes4 _selector,
+        ProxyDeployArgs memory _args,
+        string memory _contractName,
+        ExtraInstruction[] memory _instructions
+    )
+        internal
+        returns (address payable)
+    {
+        bool loadCanFail = _hasInstruction(
+            _instructions, ExtraInstruction({ key: "PermittedProxyDeployment", data: bytes(_contractName) })
+        ) || _hasInstruction(_instructions, ExtraInstruction({ key: "PermittedProxyDeployment", data: bytes("ALL") }));
+
+        // Try to load the proxy from the source.
+        (bool success, bytes memory result) = address(_source).staticcall(abi.encodePacked(_selector));
+
+        // If the load succeeded and the result is not a zero address, return the result.
+        if (success && abi.decode(result, (address)) != address(0)) {
+            return payable(abi.decode(result, (address)));
+        } else if (!loadCanFail) {
+            // Load not permitted to fail but did, revert.
+            revert OPContractsManagerV2_ProxyMustLoad();
+        }
+
+        // We've failed to load, but we allowed that failure.
+        // Deploy the right proxy depending on the contract name.
+        address ret;
+        if (LibString.eq(_contractName, "L1StandardBridge")) {
+            // L1StandardBridge is a special case ChugSplashProxy (legacy).
+            ret = Blueprint.deployFrom(
+                blueprints().l1ChugSplashProxy,
+                _computeSalt(_args.l2ChainId, _args.saltMixer, "L1StandardBridge"),
+                abi.encode(_args.proxyAdmin)
+            );
+
+            // ChugSplashProxy requires setting the proxy type on the ProxyAdmin.
+            _args.proxyAdmin.setProxyType(ret, IProxyAdmin.ProxyType.CHUGSPLASH);
+        } else if (LibString.eq(_contractName, "L1CrossDomainMessenger")) {
+            // L1CrossDomainMessenger is a special case ResolvedDelegateProxy (legacy).
+            string memory l1XdmName = "OVM_L1CrossDomainMessenger";
+            ret = Blueprint.deployFrom(
+                blueprints().resolvedDelegateProxy,
+                _computeSalt(_args.l2ChainId, _args.saltMixer, "L1CrossDomainMessenger"),
+                abi.encode(_args.addressManager, l1XdmName)
+            );
```

Comment:
Is it better to have this function be really big and detect which contracts get the nonstandard proxies, or should we break them out?

Reply:
This felt marginally cleaner to me than having three different functions. Do you have a specific alternative in mind?